File: api/crs/crs_Axpy.m
function y = crs_Axpy(A, x, y, nthreads)
%crs_Axpy Compute y=y+A*x for a sparse matrix A in CRS format,
% assuming that number of columns of A is equal to size(x,1).
%
%      y = crs_Axpy(A, x ,y)
% Computes y=A*x+y in serial.
%
%      y = crs_Axpy(A, x, y, nthreads)
% Computes y=A*x+y Testing nthreads OpenMP threads, where nthreads is an int32.
% In this mode, there are implicit barriers in the function.
%
%      y = crs_Axpy(A, x, y, [])
% Computes y=A*x+y locally within each OpenMP thread, assuming the
% parallel region has already been initialized. The argument y has
% been preallocated. In this mode, there is no implicit barrier.
%
%      y = crs_Axpy(A, x, y, nthreads, comm)
% also performs communication within the MPI communicator.
% It assumes that A is partitioned columnwise, and hence A*x is
% a partial sum on each process. An allreduce is performed on y
% within the communicator.
%
% Note that if A is partitioned rowwise, then y must be duplicated
% on all processes, and prodAx should be called without the communicator.
%
%      y = crs_Axpy(A, x, y, nthreads, comm, pgmsg [, pgsz])
% specifies an additional message that should be summed during
% allreduce and strored in y. This is to piggy-back on allreduce
% to help aggregate small messages to optimize MPI communication.
%
% The compiled version reqiures A and x to be double precision.
% The M file works for both single and double precision.

% See http://www.netlib.org/linalg/html_templates/node98.html

%#codegen -args {crs_matrix, coder.typeof(0, [inf,inf]), coder.typeof(0, [inf,inf]),int32(0)}
%#codegen crs_Axpy_ser1 -args {crs_matrix, coder.typeof(0, [inf,inf]), coder.typeof(0, [inf,inf])}

if size(y, 1) < A.nrows || size(y, 2) < size(x, 2)
    m2c_error('crs_Axpy:BufferTooSmal', 'Buffer space for output y is too small.');
end


if nargin > 3 && ~isempty(nthreads)
    %% Declare parallel region
    if ~ompGetNested && ompGetNumThreads > 1 && nthreads(1) > 1
        OMP_begin_master
        m2c_warn('crs_Axpy:NestedParallel', ...
            'You are trying to use nested parallel regions. Solution may be incorrect.');
        OMP_end_master
    end

    OMP_begin_parallel(int32(nthreads(1)));

    %% Compute y=A*x+y
    y = crs_Axpy_kernel(A.row_ptr, A.col_ind, A.val, x, int32(size(x, 1)), ...
        y, int32(size(y, 1)), A.nrows, int32(size(x, 2)), ompGetNumThreads > 1);

    %% End parallel region
    OMP_end_parallel;
elseif nargin < 4 || ompGetNumThreads == 1
    %% Compute y=A*x+y
    y = crs_Axpy_kernel(A.row_ptr, A.col_ind, A.val, x, int32(size(x, 1)), ...
        y, int32(size(y, 1)), A.nrows, int32(size(x, 2)));
else
    %% Compute y=A*x+y
    y = crs_Axpy_kernel(A.row_ptr, A.col_ind, A.val, x, int32(size(x, 1)), ...
        y, int32(size(y, 1)), A.nrows, int32(size(x, 2)), true);
end

function y = crs_Axpy_kernel(row_ptr, col_ind, val, ...
    x, x_m, y, b_m, nrows, nrhs, varargin)

coder.inline('never');
if ~isempty(varargin) && varargin{1}
    [istart, iend] = OMP_local_chunk(nrows);
else
    istart = int32(1); iend = nrows;
end

xoffset = int32(0); boffset = int32(0);
for k = 1:nrhs
    for i = istart:iend
        t = y(boffset+i);

        for j = row_ptr(i):row_ptr(i+1) - 1
            t = t + val(j) * x(xoffset+col_ind(j));
        end
        y(boffset+i) = t;
    end
    xoffset = xoffset + x_m; boffset = boffset + b_m;
end

function test %#ok<DEFNU>
%!test
%! if ~exist(['crs_Axpy.' mexext], 'file')
%!    m=100; n = 20;
%! else
%!    m=10000; n = 2000;
%! end
%! tic; sp = sprand(m,n,0.5); x=rand(size(sp,2),2);
%! fprintf(1, '\n\tGenerated random matrix in %g seconds\n', toc);
%! tic; b0 = sp*x + ones(m, 2);
%! fprintf(1, '\tComputed reference solution in %g seconds\n', toc);
%! tic; A = crs_matrix(sp);
%! fprintf(1, '\tConverted into crs_matrix in %g seconds\n', toc);
%! fprintf(1, '\tTesting serial: ');
%! tic; b1 = crs_Axpy(A, x, ones(m, 2));
%! fprintf(1, 'Done in %g seconds\n ', toc);
%! assert(norm(b0-b1)<=1.e-10);

%! b2 = zeros(size(sp,1),2);
%! for nthreads=int32([1 2 4 8])
%!     if nthreads>ompGetMaxThreads; break; end
%!     fprintf(1, '\tTesting %d thread(s): ', nthreads);
%!     tic; b2 = crs_Axpy(A, x, ones(m, 2), nthreads);
%!     fprintf(1, 'Done in %g seconds\n ', toc);
%!     assert(norm(b0-b2)<=1.e-10);
%! end

File: api/crs/codegen/lib/crs_Axpy/crs_Axpy.c
#include "crs_Axpy.h"
#include "m2c.h"
#include "omp.h"

static void b_crs_Axpy_kernel(const emxArray_int32_T *row_ptr, const
  emxArray_int32_T *col_ind, const emxArray_real_T *val, const emxArray_real_T
  *x, int x_m, emxArray_real_T *y, int b_m, int nrows, int nrhs);
static void crs_Axpy_kernel(const emxArray_int32_T *row_ptr, const
  emxArray_int32_T *col_ind, const emxArray_real_T *val, const emxArray_real_T
  *x, int x_m, emxArray_real_T *y, int b_m, int nrows, int nrhs, boolean_T
  varargin_1);
static void m2c_error(void);
static void m2c_warn(void);
static void b_crs_Axpy_kernel(const emxArray_int32_T *row_ptr, const
  emxArray_int32_T *col_ind, const emxArray_real_T *val, const emxArray_real_T
  *x, int x_m, emxArray_real_T *y, int b_m, int nrows, int nrhs)
{
  int xoffset;
  int boffset;
  int k;
  int i;
  int b_i;
  int t_tmp;
  double t;
  int c_i;
  int j;
  xoffset = -1;
  boffset = -1;
  for (k = 0; k < nrhs; k++) {
    for (i = 0; i < nrows; i++) {
      b_i = i + 1;
      t_tmp = boffset + b_i;
      t = y->data[t_tmp];
      c_i = row_ptr->data[b_i - 1];
      b_i = row_ptr->data[b_i] - 1;
      for (j = c_i; j <= b_i; j++) {
        t += val->data[j - 1] * x->data[xoffset + col_ind->data[j - 1]];
      }

      y->data[t_tmp] = t;
    }

    xoffset += x_m;
    boffset += b_m;
  }
}

static void crs_Axpy_kernel(const emxArray_int32_T *row_ptr, const
  emxArray_int32_T *col_ind, const emxArray_real_T *val, const emxArray_real_T
  *x, int x_m, emxArray_real_T *y, int b_m, int nrows, int nrhs, boolean_T
  varargin_1)
{
  int istart;
  int iend;
  int b_remainder;
  int threadID;
  int k;
  int chunk;
  int i;
  double t;
  int b_i;
  int i1;
  int j;
  if (varargin_1) {
    iend = omp_get_num_threads();
    if (iend == 1) {
      istart = 0;
      iend = nrows;
    } else {
      threadID = omp_get_thread_num();
      chunk = nrows / iend;
      b_remainder = nrows - iend * chunk;
      if (b_remainder < threadID) {
        iend = b_remainder;
      } else {
        iend = threadID;
      }

      istart = threadID * chunk + iend;
      iend = (istart + chunk) + (threadID < b_remainder);
    }
  } else {
    istart = 0;
    iend = nrows;
  }

  b_remainder = -1;
  threadID = -1;
  for (k = 0; k < nrhs; k++) {
    for (i = istart + 1; i <= iend; i++) {
      chunk = threadID + i;
      t = y->data[chunk];
      b_i = row_ptr->data[i - 1];
      i1 = row_ptr->data[i] - 1;
      for (j = b_i; j <= i1; j++) {
        t += val->data[j - 1] * x->data[b_remainder + col_ind->data[j - 1]];
      }

      y->data[chunk] = t;
    }

    b_remainder += x_m;
    threadID += b_m;
  }
}

static void m2c_error(void)
{
  M2C_error("crs_Axpy:BufferTooSmal", "Buffer space for output y is too small.");
}

static void m2c_warn(void)
{
  M2C_warn("crs_Axpy:NestedParallel",
           "You are trying to use nested parallel regions. Solution may be incorrect.");
}

void crs_Axpy(const struct0_T *A, const emxArray_real_T *x, emxArray_real_T *y,
              int nthreads)
{
  int n;
  if ((y->size[0] < A->nrows) || (y->size[1] < x->size[1])) {
    m2c_error();
  }

  n = omp_get_nested();
  if (n == 0) {
    n = omp_get_num_threads();
    if ((n > 1) && (nthreads > 1)) {

#pragma omp master
      {
        m2c_warn();
      }

    }
  }

#pragma omp parallel default(shared) num_threads(nthreads)
  {
    n = omp_get_num_threads();
    crs_Axpy_kernel(A->row_ptr, A->col_ind, A->val, x, x->size[0], y, y->size[0],
                    A->nrows, x->size[1], n > 1);
  }

}

void crs_Axpy_initialize(void)
{
}

void crs_Axpy_ser1(const struct0_T *A, const emxArray_real_T *x, emxArray_real_T
                   *b)
{
  if ((b->size[0] < A->nrows) || (b->size[1] < x->size[1])) {
    m2c_error();
  }

  b_crs_Axpy_kernel(A->row_ptr, A->col_ind, A->val, x, x->size[0], b, b->size[0],
                    A->nrows, x->size[1]);
}

void crs_Axpy_terminate(void)
{
}
